pkg/devfile/adapters/common/command.go

```diff
@@ -7,140 +7,153 @@ import (
+			supportedCommand = command
 			return supportedCommand, nil
 		}
 	}
 
-	// The command was not found
-	msg := fmt.Sprintf("The command \"%v\" was not found in the devfile", commandName)
-	if required {
-		// Not found and required, return an error
-		err = fmt.Errorf(msg)
-	} else {
-		// Not found and optional, so just log it
-		klog.V(3).Info(msg)
-	}
-
-	return
-}
+	if commandName == "" {
+		// if default command is not found return the first command found for the matching type.
+		for _, command := range data.GetCommands() {
+			if command.Exec.Group.Kind == groupType {
+				supportedCommand = command
+				return supportedCommand, nil
+			}
 
-// getSupportedCommandActions returns the supported actions for a given command and any errors
-// If some actions are supported and others have errors both the supported actions and an aggregated error will be returned.
-func getSupportedCommandActions(data data.DevfileData, command common.DevfileCommand) (supportedCommandActions []common.DevfileCommandAction, err error) {
-	klog.V(3).Infof("Validating actions for command: %v ", command.Name)
-
-	problemMsg := ""
-	for i, action := range command.Actions {
-		// Check if the command action is of type exec
-		err := validateAction(data, action)
-		if err == nil {
-			klog.V(3).Infof("Action %d maps to component %v", i+1, *action.Component)
-			supportedCommandActions = append(supportedCommandActions, action)
-		} else {
-			problemMsg += fmt.Sprintf("Problem with command \"%v\" action #%d: %v", command.Name, i+1, err)
 		}
 	}
 
-	if len(problemMsg) > 0 {
-		err = fmt.Errorf(problemMsg)
+	// if any command specified via flag is not found in devfile then it is an error.
+	if commandName != "" {
+		err = fmt.Errorf("The command \"%v\" is not found in the devfile", commandName)
+	} else {
+		msg := fmt.Sprintf("The command type \"%v\" is not found in the devfile", groupType)
+		// if run command is not found in devfile then it is an error
+		if groupType == common.RunCommandGroupType {
+			err = fmt.Errorf(msg)
+		} else {
+			klog.V(3).Info(msg)
+
+		}
 	}
 
 	return
 }
 
-// validateAction validates the given action
-// 1. action has to be of type exec
+// validateCommand validates the given command
+// 1. command has to be of type exec
 // 2. component should be present
 // 3. command should be present
-func validateAction(data data.DevfileData, action common.DevfileCommandAction) (err error) {
+func validateCommand(data data.DevfileData, command common.DevfileCommand) (err error) {
 
 	// type must be exec
-	if *action.Type != common.DevfileCommandTypeExec {
-		return fmt.Errorf("Actions must be of type \"exec\"")
+	if command.Exec == nil {
+		return fmt.Errorf("Command must be of type \"exec\"")
 	}
 
 	// component must be specified
-	if action.Component == nil || *action.Component == "" {
-		return fmt.Errorf("Actions must reference a component")
+	if command.Exec.Component == "" {
```

Comment:
I'm wondering if we should handle this as part of the devfile validator, rather than here. Then it would save a bunch of time if the user was using a devfile with invalid commands

Reply:
@johnmcollier I also think validation should happen after parsing, but refactoring validation logic is not part of this PR :)
you can raise new issue for that.

Reply:
@adisky Fair enough 😄